src/lib/log.c

```diff
@@ -20,50 +20,89 @@
 #include <stdarg.h>
 #include <stdlib.h>
 #include <string.h>
 #include <time.h>
+#include <unistd.h>
 
 #include "log.h"
+
 #define DEBUG_HEADER_SIZE 50
+#define LINE_MAX _POSIX2_LINE_MAX
 
 static void log_internal(FILE *out, const char *file, int line, const char *label, const char *format, va_list args_list);
 static int cur_log_level = LOG_INFO;
 static log_fn_t log_function = log_internal;
 
-static void log_internal(FILE *out, const char *file, int line, const char *label, const char *format, va_list args_list)
+static void print_debug_info(FILE *out, const char *file, int line)
 {
 	char time_str[50];
 	struct tm *timeinfo;
 	int printed;
+	time_t currtime;
+
+	time(&currtime);
+	timeinfo = localtime(&currtime);
+	strftime(time_str, sizeof(time_str), "%Y-%m-%d %H:%M:%S", timeinfo);
+	printed = fprintf(out, "%s (%s:%d)", time_str, file, line);
+	if (printed < DEBUG_HEADER_SIZE) {
+		fprintf(out, "%*c", 50 - printed, ' ');
+	}
+}
+
+static void log_internal(FILE *out, const char *file, int line, const char *label, const char *format, va_list args_list)
+{
+
+	char msg[LINE_MAX];
+
+	/* do not print the carriage return found at the beginning of the message if
+	 * in debug mode or if printing to a file, that would mess up the output */
+	while (strncmp(format, "\r", 1) == 0) {
+		if (cur_log_level != LOG_DEBUG && isatty(fileno(out))) {
+			fprintf(out, "%c", format[0]);
+		}
+		format++;
+	}
 
-	/* if the message starts with '\n' add a line break before the message */
-	while (strncmp(format, "\n", 1) == 0 || strncmp(format, "\r", 1) == 0) {
-		if (out == stdout || out == stderr) {
+	/* do not print the line breaks found at the beginning of the message
+	 * only if printing to a file */
+	while (strncmp(format, "\n", 1) == 0) {
+		if (isatty(fileno(out))) {
+			if (cur_log_level == LOG_DEBUG) {
+				print_debug_info(out, file, line);
+			}
 			fprintf(out, "%c", format[0]);
 		}
 		format++;
 	}
 
+	/* if we don't have anything else to print, finish here */
+	if (strcmp(format, "") == 0) {
+		return;
+	}
+
 	// In debug mode, print more information
 	if (cur_log_level == LOG_DEBUG) {
-		time_t currtime;
-		time(&currtime);
-		timeinfo = localtime(&currtime);
-		strftime(time_str, sizeof(time_str), "%Y-%m-%d %H:%M:%S", timeinfo);
-
-		printed = fprintf(out, "%s (%s:%d)", time_str, file, line);
-		if (printed < DEBUG_HEADER_SIZE) {
-			fprintf(out, "%*c", 50 - printed, ' ');
-		}
+		print_debug_info(out, file, line);
 	}
 
 	if (label) {
 		fprintf(out, "%s: ", label);
 	}
-	vfprintf(out, format, args_list);
-	fflush(out);
+
+	/* all debug messages need to finish with a line break, even if the
+	 * user didn't add one to the message, otherwise the output will be
+	 * messed up
+	 * Note that the "\n" could have been provided as an argument to the
+	 * string, as in this example: info("example%s", flag ? "" : "\n")
+	 * so we need to pass the string to a variable first to inspect it */
+	vsnprintf(msg, LINE_MAX, format, args_list);
```

Comment:
when we replaced vfprintf() with vsnprintf() + fprintf()?

Reply:
oh, I got it. to check the last \n

Reply:
I added an explanation why I did that. If we do something like this:
`info("example%s", flag ? "" : "\n")`

We need that so we can inspect the string to see if it already contains a "\n" at the end. If we remove that part we are going to be printing an extra blank line in debug mode in those cases, so it is not a big deal, if you prefer to print the extra line I can remove that and go back to using vfprintf()

Reply:
no, it's fine the way it was. I didn't get that at first